src/ocean/meta/types/ReduceType.d

```diff
@@ -0,0 +1,199 @@
+
+public template ReduceType ( T, Reducer )
+{
+    const ReduceType = ReduceTypeImpl!(Reducer).reduce!(T)();
+}
+
+version(UnitTest)
+{
+    private struct ExampleReducer
+    {
+        alias int Result;
+        const seed = 0;
+
+        static Result accumulate ( Result accum, Result next )
+        {
+            return accum + next;
+        }
+
+        static Result visit ( T ) ( )
+        {
+            if (isIntegerType!(T))
+                return 1;
+            else
+                return 0;
+        }
+    }
+}
+
+unittest
+{
+    static assert (ReduceType!(int, ExampleReducer) == 1);
+    static assert (ReduceType!(int[int], ExampleReducer) == 2);
+}
+
+/*******************************************************************************
+
+    Verifies that `Reducer` is an aggregate type that conforms type reducer
+    requirements and issues static assertion otherwise.
+
+    Params:
+        Reducer = aggregate type to check
+
+*******************************************************************************/
+
+public template assertValidReducer ( Reducer )
+{
+    static assert (
+        is(Reducer.Result),
+        Reducer.stringof ~  " must define `Result` type alias"
+    );
+    static assert (
+        is(typeof(Reducer.seed) == Reducer.Result),
+        Reducer.stringof ~ " must define `seed` constant of type `Result`"
+    );
+    static assert (
+        is(typeof(Reducer.accumulate)),
+        Reducer.stringof ~ " must define `accumluate` CTFE function"
+    );
+    static assert (
+        is(typeof(Reducer.visit!(int))),
+        Reducer.stringof ~ " must define `visit` templated function"
+    );
+
+    alias void assertValidReducer;
+}
+
+/*******************************************************************************
+
+    Implementation for `ReduceType` handling type recursion
+
+    Params:
+        Reducer = see `ReduceType` param documentation
+
+*******************************************************************************/
+
+private struct ReduceTypeImpl ( Reducer )
+{
+    // Give better error messages for wrong `Reducer` definitions
+    alias assertValidReducer!(Reducer) check;
+
+    // Helper to calculate new value and update accumulator in one step
+    static private void accumulate ( T ) ( ref T accum, T next )
+    {
+        accum = Reducer.accumulate(accum, next);
+    }
+
+    // Main recursive visiting implementation
+    static private Reducer.Result reduce ( T ) ( )
+    {
+        auto result = Reducer.seed;
+        accumulate(result, Reducer.visit!(T)());
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isAggregateType!(T))
+        {
+            foreach (TElem; typeof(T.init.tupleof))
+                accumulate(result, reduce!(TElem)());
+        }
+        else static if (isArrayType!(T))
+        {
+            static if (is(T U : U[]))
```

Comment:
Suggest using the return value of the nice new `isArrayType` template which already tells you what kind of array it is.